source/elements/oneDPL/source/parallel_api.rst

```diff
@@ -262,6 +262,19 @@ Iterators
 The oneDPL iterators are defined in the :code:`<oneapi/dpl/iterator>` header,
 in :code:`namespace oneapi::dpl`.
 
+Let us define a named requirement, :code:`ValidParallelIteratorSource`, to describe valid random access iterator-like
+types that can be used as source for oneDPL's iterators as described below.
+The type :code:`Iter` satisfies the :code:`ValidParallelIteratorSource` named requirement if it satisfies
+at least one of the following:
+
+ * :code:`Iter` is a random access iterator
+ * :code:`Iter` is the unspecified iterator-like type returned by :code:`oneapi::dpl::begin` or :code:`oneapi::dpl::end`
+ * :code:`Iter` is a valid :code:`permutation_iterator`
+ * :code:`Iter` is a valid :code:`transform_iterator`
+ * :code:`Iter` is a valid :code:`counting_iterator`
+ * :code:`Iter` is a valid :code:`discard_iterator`
+ * :code:`Iter` is a valid :code:`zip_iterator`
```

Comment:
What does "valid" indicate here? We are speaking about types, not instances.

Reply:
My intention is to indicate that if you have something like `zip_iterator<transform_iterator<bad_source, func>,counting_iterator<int>>` , where `bad_source` is does not meet the requirements that `transform_iterator` lays out, that `bad_source` iterator ruins all combinations of types based upon it.  I guess that is obvious without any word like "valid", so we can remove it.

Reply:
rephrased from your suggestion to remove valid.

Reply:
For something like you described, the requirement is violated for `transform_iterator<bad_source, func>` - which I would assume lead to a compilation error because the type cannot be instantiated.
The situation is different though if such a type can be instantiated but the behavior is wrong/undefined. Is that the case?

Reply:
Yes, I believe we could make a source type `bad_source` which would not cause compilation errors but would not satisfy the requirements set out there and the behavior would therefore be undefined.  I need to think a little to contrive such a type, but I think it exists.  

Also, there may be compile errors for `bad_source` iterators which may not be so clear, as we do not currently have some compile time check for this listing of types at the top level with a clear user facing error. It may manifest as some missing operator buried within an algorithm.

Reply:
The types of errors do not matter as it comes to the specification; it's a quality of implementation issue. However if it is not always a compilation error, that changes the things.

I see two possible approaches:
- allow this in the specification, so that it matches the current implementation; or
- disallow this in the specification (i.e. require a compile-time error and not UB), and make a "known issue" note for the implementation that it violates the spec - of course with the intention to fix it later.

My preference goes to the second option.

Reply:
Let me experiment with this a little and make sure I am not wrong.  It may be that it would result in compiler errors or correct (but unspecified) results.  For instance, I think we don't need ALL the requirements of random access iterator in practice, but it covers what we do need.

I'll try to come up with some example(s) which result in a successful build, but incorrect results by not meeting this specification.

Reply:
I have not yet tested this in practice, but lets consider an iterator type `bad_source` which is not a "random access iterator" as the c++20 standard defines it. Specifically, lets say it is defined such that the bracket operator `iter[n]` does not return something equivalent to `*(iter + n)`, but rather it just ignores `n` and returns `*iter`.  

`bad_source` still defines all the operators, functions, etc. which would be needed by a random access iterator, it just does not match the semantics required of a "random access iterator" for `operator[]`.  If this were used as input to a oneDPL algorithm, it would cause problems at runtime, but I think it would build successfully. Similarly, if inputs of `transform_iterator<bad_source, func>`, or `zip_iterator<transform_iterator<bad_source, func>, int*>`  were used as input to a oneDPL algorithm they would experience similar issues at runtime, not compile time I believe. 

I think that this is OK, because they are disobeying the spec, so their behavior is undefined.  `bad_source` is not a valid oneDPL input, nor are iterators which wrap `bad_source`.  The original point was merely that our iterators can be used as source iterators for our iterators where specified, but only when they adhere to the specification of those iterators. 

Do we need extra language to state this, or do you think this is covered by the current language?

Reply:
OK, I see now what you mean. It's about the source violating semantical requirements that cannot be checked at compile time, and so our adjusting iterators "inherit" the undefined behavior of the source. I do not think any special language is necessary for this.

Thinking of a case where our iterators can **introduce** undefined behavior even though the source was fully conformant, I find the `permutation_iterator` standing out, as both its data source and the map might be perfectly valid, but using the map values to access the data source might make that access invalid. This case may deserve some extra wording indicating a potential undefined behavior if certain conditions are not met.

Reply:
I believe that the requirements specified here for the "source data" and for the "map" of a `permutation_iterator` allow it to function appropriately and not introduce any undefined behavior as long as the requirements are met.  I will take another look through this to confirm.

We explored this in a good amount of detail with our input_data_sweep tests, and a number of bugfixes earlier this year.

Reply:
In our existing implementation of oneDPL and permutation iterator, there is an issue when using a `std::vector::iterator` (host iterator) as the source iterator for a permutation iterator.  This is because without sweeping through the whole map vector, we do not know what range of data to prepare by copying into a buffer.  Here in the spec, we don't specify the details of getting data to the device (host iterators, USM allocators, etc.), so this is not something that makes sense to be in the specification.  Otherwise, the `permutation_iterator` should be reliable as described here without introducing any undefined behavior.

This is noted in the oneDPL implementation's documentation [here](https://github.com/oneapi-src/oneDPL/blob/1266c3867ab2aa234519b8f7cffae3e95a2483bc/documentation/library_guide/parallel_api/iterators.rst?plain=1#L105).

Reply:
Thank you. This discussion is resolved then.